handlers.go

```diff
@@ -15,10 +36,132 @@ func formatIndexPage(appURL string) string {
 	return page
 }
 
-// getIndexHandler returns an index handler that informs the visitor that this
-// host runs inside an enclave.
-func getIndexHandler(cfg *Config) http.HandlerFunc {
+// rootHandler returns a handler that informs the visitor that this host runs
+// inside an enclave.  This is useful for testing.
+func rootHandler(cfg *Config) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
 		fmt.Fprintln(w, formatIndexPage(cfg.AppURL))
 	}
 }
+
+// reqSyncHandler returns a handler that lets the enclave application request
+// state synchronization, which copies the given remote enclave's state into
+// our state.
+func reqSyncHandler(e *Enclave) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		q := r.URL.Query()
+		// The 'addr' parameter must have the following form:
+		// https://example.com:8443
+		addrs, ok := q["addr"]
+		if !ok {
+			http.Error(w, errNoAddr.Error(), http.StatusBadRequest)
+			return
+		}
+		addr := addrs[0]
+
+		// Are we dealing with a well-formed URL?
+		if _, err := url.Parse(addr); err != nil {
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			return
+		}
+
+		if err := RequestKeys(addr, e.KeyMaterial); err != nil {
```

Comment:
CodeQL is making a similar point about trusting foreign input here; IIRC we assumed it as under local control before.

You've already parsed `addr` into a `url` for validation purposes. Maybe address the vulnerability warning by stripping any query items from `addr`, passing it as a url, and then use `JoinPath` inside `RequestKeys` to append the correct path before issuing the nonce and transfer requests?

Reply:
Yes, this is internal and therefore trusted input. I marked it as false positive in CodeQL. How about we address this later, when we narrowed down the sync mechanism? I made a note in https://github.com/brave/nitriding/issues/44.

Reply:
Ok.